internal/util/upgrade.go

```diff
@@ -369,13 +381,47 @@ func checkWritePermission(filePath string) error {
 
 	file, err := os.OpenFile(filePath, os.O_WRONLY, info.Mode())
 	if err != nil {
+		if strings.Contains(err.Error(), "text file busy") {
+			return fmt.Errorf("binary is currently running: %w", err)
+		}
 		return err
 	}
 	file.Close()
 
 	return nil
 }
 
+func updateRunningBinary(currentExe, newBinary string, fileMode os.FileMode) error {
+	dir := filepath.Dir(currentExe)
+	tmpBinary := filepath.Join(dir, ".agentuity.new")
+	oldBinary := filepath.Join(dir, ".agentuity.old")
+
+	if err := copyFile(newBinary, tmpBinary); err != nil {
+		return fmt.Errorf("failed to copy new binary to temp location: %w", err)
+	}
+
+	if err := os.Chmod(tmpBinary, fileMode); err != nil {
+		return fmt.Errorf("failed to set permissions on new binary: %w", err)
+	}
+
+	script := fmt.Sprintf(`#!/bin/sh
+# Wait for the process to exit
+sleep 1
+# Perform the update
+mv "%s" "%s"
+mv "%s" "%s"
+rm "%s"
+`, currentExe, oldBinary, tmpBinary, currentExe, oldBinary)
+
+	updateScript := filepath.Join(dir, ".agentuity_updater.sh")
```

Comment:
does this script get deleted? should we check for it later on startup and delete it? how can we handle it?